context.go

```diff
@@ -12,161 +12,248 @@ import (
+		return &fileContent{format: format, filename: data}
+	}
+}
+
+type inputFormat string
+
+const (
+	yamlFormat = inputFormat("yaml")
+	jsonFormat = inputFormat("json")
+	kvFormat   = inputFormat("kv")
+	textFormat = inputFormat("text")
+)
+
+// parseFormat part of content (content = :format:data)
+// nothing there -> nil
+func parseFormat(content string) (*inputFormat /* format */, string /* data */) {
+	if !strings.HasPrefix(content, ":") {
+		// it's just a text string/filename - no format specifier
+		return nil, content
+	}
+
+	content = content[1:]
+
+	// Hack: if the first thing after the ':' is a '+', it must be
+	// raw yaml/json (and we don't want to split on ':', as that might
+	// be a valid char...)
+	if strings.HasPrefix(content, "+") {
+		f := yamlFormat
+		return &f, content
+	}
+
+	splitContent := strings.SplitN(content, ":", 2)
+
+	if len(splitContent) == 1 {
+		return nil, splitContent[0]
 	}
+
+	// we don't validate here to avoid errors - will fail on eval.
+	f := inputFormat(splitContent[0])
+	return &f, splitContent[1]
 }
 
-type commonContext struct {
-	key string
-	raw bool
+type context struct {
+	original string
+	key      string
+
+	content content
 }
 
-func (cc *commonContext) cleanupResult(result interface{}) (interface{}, error) {
-	var finalResult interface{}
-	if cc.raw {
-		switch r := result.(type) {
-		case []byte:
-			finalResult = string(r)
-		default:
-			// otherwise, let's just let JSON-e complain...
-			finalResult = result
-		}
-	} else {
-		var rawBytes []byte
-		switch r := result.(type) {
-		case string:
-			rawBytes = []byte(r)
-		case []byte:
-			rawBytes = r
-		default:
-			return nil, fmt.Errorf("somehow got a non string/byte slice: %q", result)
+func (c *context) eval() (interface{}, error) {
+	result, err := c.content.load()
+	if err != nil {
+		return nil, err
+	}
+
+	if c.key != "" {
+		return map[string]interface{}{c.key: result}, nil
+	}
+
+	return result, nil
+}
+
+func loadBytes(format inputFormat, data []byte) (interface{}, error) {
+	switch format {
+	case jsonFormat:
+		var result interface{}
+		if err := json.Unmarshal(data, &result); err != nil {
+			return nil, err
 		}
-		if err := yaml_ghodss.Unmarshal(rawBytes, &finalResult); err != nil {
+		return result, nil
+	case yamlFormat:
+		var result interface{}
+		if err := yaml_ghodss.Unmarshal(data, &result); err != nil {
 			return nil, err
 		}
-	}
+		return result, nil
+	case textFormat:
+		return string(data), nil
+	case kvFormat:
+		// TODO unicode?
+		lines := strings.Split(string(data), "\n")
+		result := make(map[string]interface{}, len(lines))
+		for _, line := range lines {
+			if line == "" {
+				continue
+			}
 
-	if cc.key != "" {
-		return map[string]interface{}{cc.key: finalResult}, nil
+			splitLine := strings.SplitN(line, " ", 2)
+			if len(splitLine) != 2 {
+				return nil, fmt.Errorf("line not in kv format: %q", line)
+			}
+			result[splitLine[0]] = splitLine[1]
```

Comment:
Last key wins for duplicate key.

Reply:
yes, I should have commented. Should it error out? I can change.

Reply:
Hm, I think it should probably log to stderr about this? Maybe on debug level (disabled by default?). There should be a way to discover what is happening and debug logging is what comes to mind here.